lnwallet/musig_session.go

```diff
@@ -0,0 +1,535 @@
+}
+
+// FromSchnorrShell takes a schnorr signature and parses out the last 32 bytes
+// as a normal musig2 partial signature.
+func (p *MusigPartialSig) FromSchnorrShell(sig *schnorr.Signature) {
+	var (
+		partialS      btcec.ModNScalar
+		partialSBytes [32]byte
+	)
+	copy(partialSBytes[:], sig.Serialize()[32:])
+	partialS.SetBytes(&partialSBytes)
+
+	p.sig = &musig2.PartialSignature{
+		S: &partialS,
+	}
+}
+
+// Verify attempts to verify the partial musig2 signature using the passed
+// message and signer public key.
+//
+// NOTE: This implements the input.Signature interface.
+func (p *MusigPartialSig) Verify(msg []byte, pub *btcec.PublicKey) bool {
+	var m [32]byte
+	copy(m[:], msg)
+
+	return p.sig.Verify(
+		p.signerNonce, p.combinedNonce, p.signerKeys, pub, m,
+		musig2.WithSortedKeys(), musig2.WithBip86SignTweak(),
+	)
+}
+
+// MusigNoncePair holds the two nonces needed to sign/verify a new commitment
+// state. The signer nonce is the nonce used by the signer (remote nonce), and
+// the verification nonce, the nonce used by the verifier (local nonce).
+type MusigNoncePair struct {
+	// SigningNonce is the nonce used by the signer to sign the commitment.
+	SigningNonce musig2.Nonces
+
+	// VerificationNonce is the nonce used by the verifier to verify the
+	// commitment.
+	VerificationNonce musig2.Nonces
+}
+
+// String returns a string representation of the MusigNoncePair.
+func (n *MusigNoncePair) String() string {
+	return fmt.Sprintf("NoncePair(verification_nonce=%x, "+
+		"signing_nonce=%x)", n.VerificationNonce.PubNonce[:],
+		n.SigningNonce.PubNonce[:])
+}
+
+// MusigSession abstracts over the details of a logical musig session. A single
+// session is used for each commitment transactions. The sessions use a JIT
+// nonce style, wherein part of the session can be created using only the
+// verifier nonce. Once a new state is singed, then the signer nonce is
+// generated. similarly, the verify then uses the received signer nonce to
+// complete the session and verify the incoming signature.
+type MusigSession struct {
+	// session is the backing musig2 session. We'll use this to interact
+	// with the musig2 signer.
+	session *input.MuSig2SessionInfo
+
+	// combinedNonce is the combined nonce of all signers.
+	combinedNonce [musig2.PubNonceSize]byte
+
+	// nonces is the set of nonces that'll be used to generate/verify the
+	// next commitment.
+	nonces MusigNoncePair
+
+	// inputTxOut is the funding input.
+	inputTxOut *wire.TxOut
+
+	// signerKeys is the set of public keys of all signers.
+	signerKeys []*btcec.PublicKey
+
+	// remoteKey is the key desc of the remote key.
+	remoteKey keychain.KeyDescriptor
+
+	// localKey is the key desc of the local key.
+	localKey keychain.KeyDescriptor
+
+	// signer is the signer that'll be used to interact with the musig
+	// session.
+	signer input.MuSig2Signer
+
+	// remoteCommit tracks if this session is for the remote commitment.
+	remoteCommit bool
```

Comment:
Based on experiences with working on the channel state machine it might be good for readability to have an enum here with local and remote. To avoid translating `!remote` to "local" in your head.

Reply:
think we should use `our/their` semantics to make it clearer.

Reply:
I like the enum idea, will adopt that. I don't think we can get around the fact that local/remove doesn't always map cleanly to our/their though. As an example, we have the signing/verification nonces: if we're making the remote commitment then our local nonce is the signing nonce, but if we're verifying one, then the local nonce is now the verification nonce.